File: app/src/main/java/com/pkk/andriod/attendence/misc/MessageExtractor.java
package com.pkk.andriod.attendence.misc;

import android.content.Context;
import android.graphics.Color;

import com.pkk.andriod.attendence.activity.TeacherActivity;

import java.util.ArrayList;
import java.util.List;

public class MessageExtractor {


    private static int noofstud;
    private static int start=0;
    private static int end=0;
    private static List<String> studip;
    private static List<Boolean> stud;
    public static List<Integer> roll;

    public MessageExtractor(int start, int end) {
        this.noofstud = end - start + 1;
        this.start = start;
        this.end = end;
        roll = new ArrayList<>();
        studip = new ArrayList<>();
        stud = new ArrayList<>();
        for (int x = 0; x < noofstud; x++) {
            stud.add(false);
            roll.add(this.start+x);
            studip.add("");
        }
    }

    public static boolean checkIP(String ip) {
        if (!studip.isEmpty())
            for (int x = 0; x < noofstud; x++) {
                if (!studip.get(x).isEmpty() && studip.get(x).equals(ip)) {
                    return false;
                }
            }
        return true;
    }

    public static boolean update(int rollno, String ip, Boolean present) {
        if (!(rollno >= start && rollno <= end)) {
            for (int i = end - start+1; i < noofstud; i++)
                if (roll.get(i) == rollno) {
                    studip.set(i, ip);
                    stud.set(i, present);
                    return true;
                }
        }
        else if(rollno >= start && rollno <= end){
            studip.set(rollno-start, ip);
            stud.set(rollno-start, present);
            return true;
        }
        return false;
    }

    public static List<Integer> getattendence() {
        return roll;
    }

    public static int getstud() {
        return noofstud;
    }

    public static List<Boolean> getstatus() {
        return stud;
    }

    public static int getcolor(Boolean bool) {
        if (bool)
            return Color.rgb(90, 205, 60);
        return Color.rgb(220, 20, 60);
    }

    public static String getattendence(Boolean bool) {
        if (bool)
            return "Present";
        return "Absent";
    }

    public void addStud(Context context, int rollno) {
        if (!(rollno >= start && rollno <= end)) {
            for (int i = end - start; i < noofstud; i++)
                if (roll.get(i) == rollno) {
                    Utils.showShortToast(context, rollno+" already present for evaluation.");
                    return;
                }
        }
        else if(rollno >= start && rollno <= end){
            Utils.showShortToast(context, rollno+" already present for evaluation.");
            return;
        }
        stud.add(false);
        roll.add(rollno);
        noofstud++;
    }

    static public void setStatus(int i, Boolean s){
        stud.set(i, s);
    }

    public int getStart() {
        return start;
    }
}

File: app/src/main/java/com/pkk/android/attendance/misc/MessageExtractor.kt
package com.pkk.android.attendance.misc

import android.content.Context
import androidx.core.content.ContextCompat
import com.pkk.android.attendance.R
import com.pkk.android.attendance.models.DeviceModel
import com.pkk.android.attendance.models.StudentModel
import java.util.*

class MessageExtractor(start: Int, end: Int) {

    private var _students: MutableList<StudentModel> = ArrayList()
    val students: List<StudentModel>
        get() = _students

    fun isIPAddressUnique(ip: String): Boolean {
        if (_students.isNotEmpty()) for (x in _students.indices) {
            if (!_students[x].ipAddress.isNullOrEmpty() && _students[x].ipAddress == ip) {
                return false
            }
        }
        return true
    }

    fun validateRollNo(rollNo: Int): Boolean {
        for (x in _students.indices)
            if (_students[x].rollNo == rollNo)
                return true
        return false
    }

    fun updateRollNoDetails(rollNo: Int, ip: String?, present: Boolean, device: DeviceModel): Int {
        for (x in _students.indices) {
            val student = _students[x]
            if (student.rollNo == rollNo) {
                _students[x] = StudentModel.getInstance(
                    rollNo,
                    present,
                    ip,
                    device.deviceName,
                    device.deviceInfo,
                    0
                )
                return x
            }
        }
        return -1
    }

    fun addStud(rollNo: Int): Boolean {
        for (student in _students) {
            if (student.rollNo == rollNo) {
                return false
            }
        }
        _students.add(StudentModel.getInstance(rollNo))
        return true
    }

    fun getStatus(roll_no: Int): Boolean {
        for (student in students) if (student.rollNo == roll_no) return student.isPresent
        return false
    }

    fun setStatus(index: Int, status: Boolean) {
        val oldStudent = _students[index]
        _students[index] = StudentModel(oldStudent.id, oldStudent.rollNo, status, null)
    }

    companion object {

        fun getColour(context: Context, bool: Boolean): Int {
            return if (bool) ContextCompat.getColor(
                context,
                R.color.present_color
            ) else ContextCompat.getColor(context, R.color.absent_color)
        }

        fun getAttendance(bool: Boolean): String {
            return if (bool) "Present" else "Absent"
        }
    }

    fun clear() {
        _students.clear()
    }

    fun setSessionId(sessionId: Long) {
        for (element in _students) {
            element.sessionId = sessionId
        }
    }

    init {
        _students.clear()
        for (x in start..end) {
            _students.add(StudentModel.getInstance(x))
        }
    }
}